django_fido/views.py

```diff
@@ -1,139 +1,193 @@
 from __future__ import unicode_literals
 
+import base64
 import logging
 from abc import ABCMeta, abstractmethod
-from copy import deepcopy
 from http.client import BAD_REQUEST
-from typing import Optional
-from urllib.parse import urlsplit, urlunsplit
+from typing import Dict, Optional, Tuple
 
 from django.conf import settings
 from django.contrib.auth import authenticate, get_user_model, login
+from django.contrib.auth.base_user import AbstractBaseUser
 from django.contrib.auth.mixins import LoginRequiredMixin
 from django.contrib.auth.views import LoginView
-from django.http import JsonResponse
+from django.forms import Form
+from django.http import HttpRequest, HttpResponse, JsonResponse
 from django.shortcuts import redirect
 from django.urls import reverse_lazy
 from django.utils.encoding import force_text
 from django.utils.translation import ugettext_lazy as _
 from django.views.generic import FormView, View
-from u2flib_server import u2f
+from fido2.server import Fido2Server, RelyingParty
 
-from .constants import (AUTHENTICATION_REQUEST_SESSION_KEY, AUTHENTICATION_USER_SESSION_KEY,
-                        REGISTRATION_REQUEST_SESSION_KEY, U2F_AUTHENTICATION_REQUEST, U2F_REGISTRATION_REQUEST)
-from .forms import U2fResponseForm
-from .models import U2fDevice
+from .constants import (AUTHENTICATION_USER_SESSION_KEY, FIDO2_AUTHENTICATION_REQUEST, FIDO2_REGISTRATION_REQUEST,
+                        FIDO2_REQUEST_SESSION_KEY)
+from .forms import Fido2AuthenticationForm, Fido2RegistrationForm
+from .models import Authenticator
 
 _LOGGER = logging.getLogger(__name__)
 
 
-class BaseU2fRequestView(View, metaclass=ABCMeta):
-    """Base view for U2F request views.
+class Fido2ViewMixin(object):
+    """
+    Mixin with common methods for all FIDO 2 views.
 
-    @cvar session_key: Session key where the U2F request is stored.
-    @cvar u2f_request_factory: Function which accepts `app_id` and `registered_keys` and returns U2F request.
+    @cvar rp_name: Name of the relying party. If None, the RP ID is used instead.
+    @cvar session_key: Session key where the FIDO 2 state is stored.
     """
 
-    session_key = None  # type: Optional[str]
-    u2f_request_factory = None  # type: Optional[staticmethod]
+    rp_name = None  # type: Optional[str]
+    session_key = FIDO2_REQUEST_SESSION_KEY
+
+    def get_rp_id(self) -> str:
+        """Return RP id - only a hostname for web services."""
+        # `partition()` is faster than `split()`
+        return self.request.get_host().partition(':')[0]  # type: ignore
 
-    def get_app_id(self):
-        """Return appId - base URL to the web."""
-        # Correct `appId` for the web application doesn't have a path.
-        # See https://developers.yubico.com/U2F/App_ID.html
-        chunks = urlsplit(self.request.build_absolute_uri('/'))
-        return urlunsplit((chunks.scheme, chunks.netloc, '', '', ''))
+    @property
+    def server(self) -> Fido2Server:
```

Comment:
Couldn't this be a cached property?

Reply:
It can, but every view uses it exactly once.

Reply:
OK.